docs/TESTPLAN.md

```diff
@@ -0,0 +1,204 @@
+percentage of the project has any unit tests.
+
+### Functional Tests
+
+Functional tests make up the majority of tests in the project. They test how a task and all of its components come together
+to work. In general, the point of functional tests is to make sure all of the units work together, and in this project they
+make sure that each Task is interacting properly with the service clients and settings. This is accomplished by passing in
+carefully crafted settings, and mocked implementations of service clients.
+
+For small tasks, the functional tests have great branch coverage without being too complicated, but for larger tasks like
+Beanstalk and CloudFormation, The tests are much longer. Therefore, it is preferable to grow the number of unit
+tests over the number of functional tests as they currently serve the purpose of both making sure units work and function well
+together.
+
+Looking back at past releases that had issues, the second most common reason for issues was bad handling of parameters. The
+functional tests are great at catching this category of problem.
+
+All of the functional tests follow the pattern:
+
+1. Create a `TaskParameters` object for the task
+2. Modify `TaskParameters` to make it go through the desired code path
+3. Create mocked service clients
+4. Run `TaskName.execute()`
+5. Verify service calls/failures on tasks that are supposed to fail
+
+#### Testing Gaps
+
+The following modules do not have any functional tests:
+
+-   CloudFormation Create Or Update
+-   Beanstalk Create Application
+
+The following modules do not have functional tests, but it would make no sense to add them:
+
+-   AWS Powershell
+-   AWS CLI
+
+The other major gap is most functional tests only have ~50% branch coverage. This can be improved by improving our unit tests
+instead.
+
+### End to End Tests
+
+End to end tests serve the purpose of making sure the tasks work when actually connecting to AWS. The goal of them is not to
+test every code path, but to make sure we can:
+
+-   Connect to AWS at all/create clients that work
+-   Instance and fixed credentials both work
+-   Any task that does heavy build/filesystem work is tested more heavily here (Lambda, Cloudformation, and beanstalk tasks)
+-   Anything that relies on the real environment is tested (like the aws shell script task or the powershell scripts)
+
+With this in mind, the goals for functional tests are: to run every task (to see if they work at all), test instance
+credentials and static credentials at least once, and to test all of as many code paths of tasks like aws shell script as
+possible.
+
+Looking back at past releases that had issues, the most common issue was a task not running at all, so
+the first goal is by far the most important.
+
+Here is the current list of tasks and how functional tests currently test them, gaps are in **bold** (unlike the previous sections
```

Comment:
1. 

> "Here is the current list of tasks and how **functional tests** ..."

Does this list apply to the functional tests (prev section) or end to end tests (this section)?
Assuming it belongs here, I suggest putting it in a subsection to "End to End Tests" called "Overview"

2. 
>  (unlike the previous sections in a separate section)

Consider removing that text.

3. The outline is great! For the gaps you've identified in bold, would it make sense to write up issues and link to them from here from each line?

Reply:
1. done
2. done
3. I don't think it's necessarily useful to have issues for all of these, these kind of "issues" I would file under continuous improvement otherwise we would always be sitting on 40 test issue tasks. None of them are untested, they could just be improved.